File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/shared/BlazeAsyncBridgeModule.kt
package com.blaze.flutterblazesdk.shared

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.Keep
import com.blaze.flutterblazesdk.utils.parsing.gson.BlazeFlutterJsonParsingError
import com.blaze.flutterblazesdk.utils.parsing.gson.parseJsonWithDetailedErrors
import com.blaze.flutterblazesdk.utils.parsing.gson.toJsonStringWithException
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.CancellableContinuation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout

/**
 * Public interface for the Blaze Async Bridge Provides async communication between native code and
 * Dart
 */
@Keep
interface BlazeAsyncBridge {

    companion object {

        /**
         * Register the async bridge module and set it as the global instance
         * @param binaryMessenger The Flutter binary messenger
         */
        internal fun registerModule(binaryMessenger: BinaryMessenger) {
            val instance = BlazeAsyncBridgeModule(binaryMessenger)
            BlazeAsyncBridgeProvider.setInstance(instance)
        }

        /**
         * Get the global async bridge instance
         * @return The async bridge instance or null if not initialized
         */
        @Keep
        fun getInstance(): BlazeAsyncBridge? {
            return BlazeAsyncBridgeProvider.instance
        }
    }

    @Keep
    sealed class TimeoutType {

        @Keep object DefaultTimeout : TimeoutType()

        @Keep object NoTimeout : TimeoutType()

        @Keep data class Seconds(val seconds: Double) : TimeoutType()

        internal val timeoutInMillis: Long
            get() =
                    when (this) {
                        is DefaultTimeout -> 2000L // Default timeout
                        is NoTimeout -> Long.MAX_VALUE // Use Long.MAX_VALUE to indicate no timeout
                        is Seconds -> (seconds * 1000).toLong() // Convert seconds to milliseconds
                    }
    }
}

/// Private empty parameters data class for methods with no parameters.
private object EmptyParams

/**
 * Call a Dart method with parameters and return the result as type T.
 *
 * @param name The name of the Dart method to call.
 * @param params The parameters to pass to the Dart method.
 * @param timeout The timeout for the method call.
 */
@Keep
@JvmOverloads
suspend inline fun <reified T, reified P> BlazeAsyncBridge.callDartMethod(
        name: String,
        params: P,
        timeout: BlazeAsyncBridge.TimeoutType = BlazeAsyncBridge.TimeoutType.DefaultTimeout
): T {
    // Cast to concrete implementation and call internal method directly
    val bridge =
            this as? BlazeAsyncBridgeModule
                    ?: throw IllegalStateException(
                            "BlazeAsyncBridge must be implemented by BlazeAsyncBridgeModule"
                    )
    val result: String? =
            bridge.callDartMethodInternal(name = name, params = params, timeout = timeout)

    // Deserialize JSON string to target type - parseJsonWithDetailedErrors handles null case
    return try {
        result.parseJsonWithDetailedErrors<T>(methodName = name)
    } catch (e: BlazeFlutterJsonParsingError) {
        throw BlazeAsyncBridgeModule.BlazeAsyncError.TypeError(
                "Dart method '$name' parsing failed: ${e.message}"
        )
    }
}

/**
 * Call a Dart method with no parameters and return the result as type T.
 *
 * @param name The name of the Dart method to call.
 * @param timeout The timeout for the Dart method call.
 */
@Keep
@JvmOverloads
suspend inline fun <reified T> BlazeAsyncBridge.callDartMethod(
        name: String,
        timeout: BlazeAsyncBridge.TimeoutType = BlazeAsyncBridge.TimeoutType.DefaultTimeout
): T {
    // Use EmptyParams internally for methods with no parameters
    return callDartMethod(name = name, params = EmptyParams, timeout = timeout)
}

/**
 * Send a fire-and-forget event to Dart (no response expected) Useful for notifications and events
 * from non-suspending contexts
 *
 * @param name The name of the Dart method to call
 * @param params The parameters to pass to the Dart method
 */
@Keep
@JvmOverloads
inline fun <reified P> BlazeAsyncBridge.sendEvent(name: String, params: P) {
    CoroutineScope(Dispatchers.Main).launch {
        try {
            // Call as nullable Unit return type (fire-and-forget)
            this@sendEvent.callDartMethod<Unit?, P>(
                    name = name,
                    params = params,
                    timeout = BlazeAsyncBridge.TimeoutType.DefaultTimeout
            )
        } catch (e: Exception) {
            Log.e("BlazeAsyncBridge", "Error sending event $name: ${e.message}")
        }
    }
}

@Keep
@PublishedApi
internal class BlazeAsyncBridgeModule(private val binaryMessenger: BinaryMessenger) :
        BlazeAsyncBridge {

    companion object {
        const val NAME = "BlazeAsyncBridge"
        private const val BRIDGE_ASYNC_FUNCTION_NAME = "BlazeAsyncBridge.dartRequest"
        private const val CHANNEL_NAME = "blaze-async-bridge"
    }

    // Thread-safe maps for callbacks
    private val callbackMap = ConcurrentHashMap<String, CompletionHandler>()

    // Main thread handler for UI operations
    private val mainHandler = Handler(Looper.getMainLooper())

    // Method channel for communication with Dart
    private val methodChannel = MethodChannel(binaryMessenger, CHANNEL_NAME)

    // Completion handler interface
    private fun interface CompletionHandler {
        fun invoke(result: String?, error: Throwable?)
    }

    // Error types
    sealed class BlazeAsyncError(message: String) : Exception(message) {
        class TimeoutError(methodName: String) :
                BlazeAsyncError("Dart method call timed out: $methodName")
        class DartError(errorMessage: String) : BlazeAsyncError(errorMessage)
        class TypeError(message: String) : BlazeAsyncError(message)
    }

    init {
        // Set up method call handler for responses from Dart
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "resolveDartResponse" -> {
                    resolveDartResponse(call.arguments as Map<String, Any>)
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
    }

    /** Internal method that returns JSON string - core implementation */
    @PublishedApi
    internal suspend fun <P> callDartMethodInternal(
            name: String,
            params: P,
            timeout: BlazeAsyncBridge.TimeoutType
    ): String? {
        val timeoutMillis = timeout.timeoutInMillis
        val callbackId = UUID.randomUUID().toString()
        try {
            return safeWithTimeout(timeoutMillis) {
                withContext(Dispatchers.Main) {
                    suspendCancellableCoroutine { continuation ->
                        try {
                            storeCallback(callbackId, continuation)
                            sendDartRequest(name, params, callbackId)
                            continuation.invokeOnCancellation { cause ->
                                cleanupCallback(callbackId)
                                Log.d(
                                        NAME,
                                        "Task cancelled for Dart method call: $name, reason: ${cause?.message}"
                                )
                            }
                        } catch (e: Throwable) {
                            cleanupCallback(callbackId)
                            Log.e(NAME, "Error setting up Dart method call: $name", e)
                            continuation.resumeWithException(e)
                        }
                    }
                }
            }
        } catch (e: TimeoutCancellationException) {
            cleanupCallback(callbackId)
            throw BlazeAsyncError.TimeoutError(name)
        } catch (e: Throwable) {
            cleanupCallback(callbackId)
            throw e
        }
    }

    // MARK: - Private Helper Methods

    private fun storeCallback(callbackId: String, continuation: CancellableContinuation<String?>) {
        val completionHandler = CompletionHandler { result, error ->
            handleCallbackResponse(callbackId, result, error, continuation)
        }
        callbackMap[callbackId] = completionHandler
    }

    private fun handleCallbackResponse(
            callbackId: String,
            result: String?,
            error: Throwable?,
            continuation: CancellableContinuation<String?>
    ) {
        // Resume continuation
        if (error != null) {
            continuation.resumeWithException(error)
        } else {
            continuation.resume(result)
        }

        // Remove callback from map
        cleanupCallback(callbackId)
    }

    // Send Dart request
    private fun <P> sendDartRequest(name: String, params: P, callbackId: String) {
        // Serialize params to JSON string using GsonUtils
        val paramsJson =
                try {
                    when (params) {
                        null -> "null"
                        is EmptyParams -> "null"
                        else -> params.toJsonStringWithException()
                    }
                } catch (e: Exception) {
                    Log.e(NAME, "Failed to serialize params to JSON for method: $name", e)
                    throw BlazeAsyncError.TypeError("Failed to serialize params: ${e.message}")
                }

        val eventParams =
                mapOf(
                        "methodName" to name,
                        "params" to paramsJson, // Send as JSON string instead of Map
                        "callbackId" to callbackId
                )

        // Invoke method on Dart side - ensure we're on main thread
        if (Looper.myLooper() == Looper.getMainLooper()) {
            // Already on main thread
            methodChannel.invokeMethod(BRIDGE_ASYNC_FUNCTION_NAME, eventParams)
        } else {
            // Switch to main thread
            Handler(Looper.getMainLooper()).post {
                methodChannel.invokeMethod(BRIDGE_ASYNC_FUNCTION_NAME, eventParams)
            }
        }
        Log.d(NAME, "Sending Dart request with JSON params: $name, callbackId: $callbackId")
    }

    /** Cleanup callback - used for cancellation */
    private fun cleanupCallback(callbackId: String) {
        callbackMap.remove(callbackId)
    }

    /**
     * Method called by Dart to resolve/reject a native call
     * resolveJSResponse
     */
    private fun resolveDartResponse(response: Map<String, Any>) {
        // Run on main thread
        mainHandler.post { handleDartResponse(response) }
    }

    // Handle Dart response
    private fun handleDartResponse(response: Map<String, Any>) {
        val callbackId = response["callbackId"] as? String
        if (callbackId == null) {
            Log.e(NAME, "Invalid response format: missing callbackId")
            return
        }

        val callback = callbackMap[callbackId]
        if (callback == null) {
            // Callback might have already been handled (e.g., due to timeout or cancellation)
            Log.d(
                    NAME,
                    "Received response for unknown callbackId: $callbackId (likely already handled)"
            )
            return
        }

        val success = response["success"] as? Boolean ?: false

        if (success) {
            // Extract JSON string from response - Dart should send data as JSON string
            val jsonString = response["data"] as? String
            callback.invoke(jsonString, null)
            Log.d(NAME, "Resolved Dart method call: callbackId: $callbackId")
        } else {
            val errorMessage = response["errorMessage"] as? String ?: "Unknown error"
            val error = BlazeAsyncError.DartError(errorMessage)
            callback.invoke(null, error)
            Log.d(NAME, "Rejected Dart method call: callbackId: $callbackId, error: $errorMessage")
        }

        // Clean up callback
        cleanupCallback(callbackId)
    }
}

internal suspend fun <T> safeWithTimeout(timeoutMs: Long, block: suspend () -> T): T {
    return if (timeoutMs != Long.MAX_VALUE) {
        withTimeout(timeoutMs) { block() }
    } else {
        block() // No timeout
    }
}

File: blaze_flutter_sdk/ios/blaze_flutter_sdk/Sources/Shared/BlazeAsyncBridgeModule.swift
import Foundation
import Flutter

public enum BlazeAsyncBridgeTimeout {
    case defaultTimeout
    case noTimeout
    case seconds(_ seconds: Double)
    
    fileprivate var timeoutInSeconds: Double {
        switch self {
        case .defaultTimeout:
            return 2 // Default timeout
        case .noTimeout:
            return Double.infinity // Use Double.infinity to indicate no timeout
        case .seconds(seconds: let seconds):
            return seconds
        }
    }
}

/// Global instance provider for BlazeAsyncBridge - matches Android pattern
private class BlazeAsyncBridgeProvider {
    static var instance: BlazeAsyncBridge?
    
    static func setInstance(_ bridge: BlazeAsyncBridge) {
        instance = bridge
    }
    
    static func getInstance() -> BlazeAsyncBridge? {
        return instance
    }
}

/// Public protocol for the Blaze Async Bridge
/// Provides async communication between native code and Dart
public protocol BlazeAsyncBridge {

}

/// Static methods for BlazeAsyncBridge - matches Android companion object pattern
extension BlazeAsyncBridge {
    
    /// Register the async bridge module and set it as the global instance
    /// - Parameter binaryMessenger: The Flutter binary messenger
    static func registerModule(messenger: FlutterBinaryMessenger) {
        let instance = BlazeAsyncBridgeModule(binaryMessenger: messenger)
        BlazeAsyncBridgeProvider.setInstance(instance)
    }
    
    public static func getInstance() -> BlazeAsyncBridge? {
        return BlazeAsyncBridgeProvider.getInstance()
    }
}

/// Private empty parameters struct for methods with no parameters
private struct EmptyCodable: Codable {}

/// Public constants and convenience methods for the Blaze Async Bridge
public extension BlazeAsyncBridge {
    
    /// Generic method that accepts Encodable parameters and returns the specific Decodable type
    /// This method automatically serializes the params to JSON and infers the return type
    func callDartMethod<T: Decodable, P: Encodable>(_ name: String,
                                                    params: P,
                                                    timeout: BlazeAsyncBridgeTimeout = .defaultTimeout) async throws -> T {
        guard let self = self as? BlazeAsyncBridgeModule else {
            // Should never reach here.
            throw BlazeAsyncBridgeModule.Errors.typeError(errorMessage: "BlazeAsyncBridge must be implemented by BlazeAsyncBridgeModule")
        }
        
        let jsonString = try await self.callDartMethodInternal(name,
                                                               params: params,
                                                               timeout: timeout) ?? "null"
        
        do {
            
            // Handle the case where jsonString is nil (empty response)
            guard let jsonData = jsonString.data(using: .utf8) else {
                throw BlazeAsyncBridgeModule.Errors.typeError(errorMessage: "Failed to convert JSON string to data")
            }
            
            let decoder = JSONDecoder()
            let result = try decoder.decode(T.self, from: jsonData)
            return result
            
        } catch {
            throw BlazeAsyncBridgeModule.Errors.typeError(errorMessage: "Dart method '\(name)' parsing failed: \(error)")
        }
    }
    
    func callDartMethod<T: Decodable>(_ name: String,
                                      timeout: BlazeAsyncBridgeTimeout = .defaultTimeout) async throws -> T {
        return try await callDartMethod(name,
                                        params: EmptyCodable(),
                                        timeout: timeout)
    }
    
    func callDartMethod(_ name: String,
                        timeout: BlazeAsyncBridgeTimeout = .defaultTimeout) async throws {
        let _: EmptyCodable? = try await callDartMethod(name,
                                                        params: EmptyCodable(),
                                                        timeout: timeout)
    }
    
    func callDartMethod<P: Encodable>(_ name: String,
                                      params: P,
                                      timeout: BlazeAsyncBridgeTimeout = .defaultTimeout) async throws {
        let _: EmptyCodable? = try await callDartMethod(name,
                                                        params: params,
                                                        timeout: timeout)
    }
    
    /// Send a fire-and-forget event to Dart (no response expected)
    /// Useful for notifications and events from non-async contexts
    ///
    /// - Parameters:
    ///   - name: The name of the Dart method to call
    ///   - params: The parameters to pass to the Dart method
    func sendEvent<P: Encodable>(_ name: String,
                                 params: P) {
        Task {
            do {
                 // Fire-and-forget call (nullable return value for null results)
                 try await self.callDartMethod(name, params: params, timeout: .defaultTimeout)
            } catch {
                print("BlazeAsyncBridge: Error sending event \(name): \(error)")
            }
        }
    }
    
    /// Send a fire-and-forget event to Dart with no parameters
    func sendEvent(_ name: String) {
        sendEvent(name, params: EmptyCodable())
    }

}

public class BlazeAsyncBridgeModule: NSObject, BlazeAsyncBridge {
    
    internal enum Constants {
        static let bridgeAsyncFunctionName = "BlazeAsyncBridge.dartRequest"
        static let channelName = "blaze-async-bridge"
    }
    
    enum Errors: Error {
        case timeoutError(errorMessage: String)
        case dartError(errorMessage: String)
        case typeError(errorMessage: String)
    }
    
    struct CallbackModel {
        let callbackAction: (Any?, Error?) -> Void
        let continuation: CheckedContinuation<String?, Error>
    }
    
    private var callbackMap = [String: CallbackModel]()
    
    // Method channel for communication with Dart
    private let methodChannel: FlutterMethodChannel
    
    public init(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(
            name: Constants.channelName,
            binaryMessenger: binaryMessenger
        )
        
        super.init()
        
        // Set up method call handler
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handleMethodCall(call: call, result: result)
        }
    }
    
    // Handle method calls from Dart
    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "resolveDartResponse":
            if let arguments = call.arguments as? [String: Any] {
                Task { [weak self] in
                    await self?.handleDartResponse(response: arguments)
                }
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // Replace setupTimeout with performAsyncTaskWithTimeout
    fileprivate func callDartMethodInternal<P: Encodable>(_ name: String,
                                                          params: P,
                                                          timeout: BlazeAsyncBridgeTimeout) async throws -> String? {
        let callbackId = UUID().uuidString
        
        // Determine the effective timeout
        let effectiveTimeout = timeout.timeoutInSeconds
        
        do {
            return try await withTimeout(effectiveTimeout) {
                try await withTaskCancellationHandler(operation: {
                    try await withCheckedThrowingContinuation { continuation in
                        Task { [weak self] in
                            guard let self = self else { return }
                            
                            // Store callback handling
                            await self.storeCallback(callbackId: callbackId,
                                                     continuation: continuation)
                            
                            // Send request to Dart
                            try self.sendDartRequest(name: name,
                                                     params: params,
                                                     callbackId: callbackId)
                        }
                    }
                }, onCancel: { [weak self] in
                    Task { [weak self] in
                        guard let self = self else { return }
                        
                        // Handle all cancellation (timeout + external) - resume continuation and cleanup
                        await self.resumeWithCancellation(callbackId: callbackId)
                        print("Operation cancelled for Dart method call: \(name)")
                    }
                })
            }
        } catch WithTimeoutError.timeoutError {
            let timeoutError = Errors.timeoutError(errorMessage: "Dart method call timed out: \(name)")
            await removeCallback(callbackId: callbackId)
            
            throw timeoutError
        } catch {
            // Handle other errors (including `CancellationError`).
            await removeCallback(callbackId: callbackId)
            throw error
        }
    }
    
    // MARK: - Private Helper Methods
    
    @MainActor
    private func resumeWithCancellation(callbackId: String) {
        guard let callback = callbackMap[callbackId] else {
            return
        }
        
        // Resume continuation with cancellation error if it exists
        callback.continuation.resume(throwing: CancellationError())
        
        // Remove callback from both maps
        removeCallback(callbackId: callbackId)
    }
    
    @MainActor
    private func removeCallback(callbackId: String) {
        callbackMap.removeValue(forKey: callbackId)
    }
    
    @MainActor
    private func storeCallback(callbackId: String,
                               continuation: CheckedContinuation<String?, Error>) {
        callbackMap[callbackId] = .init(
            callbackAction: { [weak self] (result, error) in
                guard let self = self else { return }
                
                self.handleCallbackResponse(callbackId: callbackId,
                                            result: result,
                                            error: error,
                                            continuation: continuation)
            },
            continuation: continuation)
    }
    
    @MainActor
    private func handleCallbackResponse(callbackId: String,
                                        result: Any?,
                                        error: Error?,
                                        continuation: CheckedContinuation<String?, Error>) {
        // Resume continuation
        if let error = error {
            continuation.resume(throwing: error)
        } else {
            // Return JSON string directly - result should already be a JSON string from Dart
            let jsonString = result as? String
            continuation.resume(returning: jsonString)
        }
        
        // Remove callback from map
        removeCallback(callbackId: callbackId)
    }
    
    private func sendDartRequest<P: Encodable>(name: String,
                                               params: P,
                                               callbackId: String) throws {
        // Serialize params to JSON string using the existing helper
        let paramsJson: String
        if params is EmptyCodable {
            paramsJson = "null"
        } else if let jsonString = params.asJsonString {
            paramsJson = jsonString
        } else {
            throw Errors.typeError(errorMessage: "Failed to serialize params to JSON string")
        }
        
        let arguments: [String: Any] = [
            "methodName": name,
            "params": paramsJson, // Send as JSON string instead of dictionary
            "callbackId": callbackId
        ]
        
        // Ensure we're on the main thread to avoid Flutter channel threading warnings
        DispatchQueue.main.async { [weak self] in
            self?.methodChannel.invokeMethod(Constants.bridgeAsyncFunctionName, arguments: arguments)
        }
    }
    
    // Method called by Dart to resolve/reject a native call
    @MainActor
    private func handleDartResponse(response: [String: Any]) {
        
        guard let callbackId = response["callbackId"] as? String else {
            return
        }
        
        guard let callback = callbackMap[callbackId] else {
            // Callback might have already been handled (e.g., due to timeout or cancellation)
            return
        }
        
        let success = response["success"] as? Bool ?? false
        
        if success {
            // Extract JSON string from response - Dart should send data as JSON string
            let jsonString = response["data"] as? String
            callback.callbackAction(jsonString, nil)
        } else {
            let errorMessage = response["errorMessage"] as? String ?? "Unknown error"
            let error = Errors.dartError(errorMessage: errorMessage)
            callback.callbackAction(nil, error)
        }
    }
}
